### ISSUE TITLE
8-byte loop not vectorized on x86 by SLP vectorizer after unrolling (cost-model issue?)

### ISSUE THREAD

**Issue by davidbolvansky** [llvm/llvm-project#46835]
|  |  |
| --- | --- |
| Bugzilla Link | [47491](https://llvm.org/bz47491) |
| Version | trunk |
| OS | Windows NT |
| CC | @alexey-bataev,@topperc,@fhahn,@RKSimon,@rotateright |

## Extended Description 
```
void pr(char* restrict r, char* restrict a){
    #pragma clang loop vectorize(enable)
    for (int i = 0; i < 8; i++){
        r[i] += a[i];
    }
}
```
Clang -O3:
```
pr:                                     # @pr
        mov     al, byte ptr [rsi]
        add     byte ptr [rdi], al
        mov     al, byte ptr [rsi + 1]
        add     byte ptr [rdi + 1], al
        mov     al, byte ptr [rsi + 2]
        add     byte ptr [rdi + 2], al
        mov     al, byte ptr [rsi + 3]
        add     byte ptr [rdi + 3], al
        mov     al, byte ptr [rsi + 4]
        add     byte ptr [rdi + 4], al
        mov     al, byte ptr [rsi + 5]
        add     byte ptr [rdi + 5], al
        mov     al, byte ptr [rsi + 6]
        add     byte ptr [rdi + 6], al
        mov     al, byte ptr [rsi + 7]
        add     byte ptr [rdi + 7], al
        ret
```
ICC or GCC -O3 (without any pragma):
```
pr:
        movq      xmm1, QWORD PTR [rdi]                         #3.9
        movq      xmm0, QWORD PTR [rsi]                         #3.17
        paddb     xmm1, xmm0                                    #3.9
        movq      QWORD PTR [rdi], xmm1                         #3.9
        ret
```

**Comment by RKSimon** [comment]
This should get addressed by https://reviews.llvm.org/D124284

**Comment by davidbolvansky** [comment]
Yeah, cool.

**Comment by RKSimon** [comment]
Fixed in 9dc4ced204d1e918d4c8d3279e52197e9a5abc94
